src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsChemicalHandler.java

```diff
@@ -0,0 +1,50 @@
+package de.srendi.advancedperipherals.common.addons.refinedstorage;
+
+import com.refinedmods.refinedstorage.api.core.Action;
+import com.refinedmods.refinedstorage.api.network.Network;
+import com.refinedmods.refinedstorage.api.network.storage.StorageNetworkComponent;
+import com.refinedmods.refinedstorage.api.storage.Actor;
+import com.refinedmods.refinedstorage.mekanism.ChemicalResource;
+import de.srendi.advancedperipherals.AdvancedPeripherals;
+import de.srendi.advancedperipherals.common.util.inventory.ChemicalFilter;
+import de.srendi.advancedperipherals.common.util.inventory.IStorageSystemChemicalHandler;
+import mekanism.api.chemical.ChemicalStack;
+import org.jetbrains.annotations.NotNull;
+
+public class RsChemicalHandler implements IStorageSystemChemicalHandler {
+
+    @NotNull
+    private final Network network;
+    private final StorageNetworkComponent component;
+
+    public RsChemicalHandler(@NotNull Network network) {
+        this.network = network;
+        this.component = network.getComponent(StorageNetworkComponent.class);
+    }
+
+    @NotNull
+    @Override
+    public ChemicalStack insertChemical(int tank, ChemicalStack resource, mekanism.api.Action action) {
+        if (resource.isEmpty())
+            return resource;
+
+        ChemicalStack chemical = resource.copy();
+        chemical.setAmount(resource.getAmount() - component.insert(ChemicalResource.ofChemicalStack(chemical), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY));
```

Comment:
I recommend extract the inserted component here for better readability. same for below